h2o-k8s/src/main/java/water/k8s/KubernetesEmbeddedConfigProvider.java

```diff
@@ -0,0 +1,94 @@
+package water.k8s;
+
+import water.init.AbstractEmbeddedH2OConfig;
+import water.init.EmbeddedConfigProvider;
+import water.k8s.lookup.KubernetesDnsLookup;
+import water.k8s.lookup.KubernetesLookup;
+import water.k8s.lookup.LookupConstraintsBuilder;
+import water.util.Log;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * A configuration provider for H2O running in Kubernetes cluster. It is able to detected H2O is being ran in K8S
+ * environment, otherwise remains inactive.
+ * <p>
+ * Uses potentially multiple strategies to discover H2O Pods on a Kubernetes cluster.
+ */
+public class KubernetesEmbeddedConfigProvider implements EmbeddedConfigProvider {
+
+    private static final String K8S_NODE_LOOKUP_TIMEOUT_KEY = "H2O_NODE_LOOKUP_TIMEOUT";
+    private static final String K8S_DESIRED_CLUSTER_SIZE_KEY = "H2O_NODE_EXPECTED_COUNT";
+
+    private boolean runningOnKubernetes = false;
+    private KubernetesEmbeddedConfig kubernetesEmbeddedConfig;
+
+    /**
+     * @return A Set
+     */
+    private static final Optional<Set<String>> resolveNodeIPs() {
```

Comment:
QQ: when we say `resolveNodeIPs` do we mean the internal ip address of each pod spun up for the h2o cluster in kubernetes or the actual physical kubernetes nodes. From a naming perspective this is potentially incorrect in the instance of the former, and could lead to confusion

Reply:
Yes, I'll rename it. The exact answer is: It only discovers pod IPs listed by the hadless service created. And the headless service only filters out (at least it should) pods with H2O. Therefore, this only discovers H2O nodes.